pkg/serviceaccount/openidmetadata.go

```diff
@@ -44,26 +47,68 @@ const (
 	JWKSPath = "/openid/v1/jwks"
 )
 
-// OpenIDMetadata contains the pre-rendered responses for OIDC discovery endpoints.
-type OpenIDMetadata struct {
-	ConfigJSON       []byte
-	PublicKeysetJSON []byte
+// OpenIDMetadataProvider returns pre-rendered responses for OIDC discovery endpoints.
+type OpenIDMetadataProvider interface {
+	GetConfigJSON() (json []byte, maxAge int, err error)
+	GetKeysetJSON() (json []byte, maxAge int, err error)
+}
+
+type openidConfigProvider struct {
+	issuerURL, jwksURI string
+	pubKeyGetter       PublicKeyGetter
+	config             atomic.Pointer[openidConfig]
+}
+type openidConfig struct {
+	configJSON []byte
+	keysetJSON []byte
 }
 
-// NewOpenIDMetadata returns the pre-rendered JSON responses for the OIDC discovery
+func (p *openidConfigProvider) GetConfigJSON() ([]byte, int, error) {
+	return p.config.Load().configJSON, p.pubKeyGetter.GetCacheAgeMaxSeconds(), nil
+}
+func (p *openidConfigProvider) GetKeysetJSON() ([]byte, int, error) {
+	return p.config.Load().keysetJSON, p.pubKeyGetter.GetCacheAgeMaxSeconds(), nil
+}
```

Comment:
When this does become dynamic, the output of these two methods can be out of sync, so a client that fetches the config JSON could get a keyset JSON that does not match.  Maybe that race does not matter in practice, but doesn't seem great.  The only way I can think to fix that is to make the `jwksURI` dynamic based on the contents of the keyset (would need wildcard RBAC on a subpath to make that work but I am sure people have hard coded the JWKs endpoint in their configs).

Reply:
> a client that fetches the config JSON could get a keyset JSON that does not match

the only way the keys inform the config json is in the list of algorithms... I don't expect the algorithms to change and I don't think it's worth the complexity of different jwk subpaths